Feature: Classification versioning

ehaivala opened the pull request:
Make classifications versionable the same way functions are. This is done because updating classification some time after it's creation to match newer functions will render the old function versions invalid since the classification data has been changed from what it used to be when the old function version was created.

Resolves #268

tuomas777 commented on metarecord/admin/classification.py:
```diff
@@ -1,10 +1,47 @@
+            raise HttpResponseNotAllowed
+
+        if not request.user.has_perm(Classification.CAN_EDIT):
+            raise HttpResponseForbidden
```
Hmm, can you actually raise `HttpResponseNotAllowed` and `HttpResponseForbidden`, I don't think they are exceptions?

ehaivala replied:
Yeah, you're correct. Good catch 👍

tuomas777 commented on locale/fi/LC_MESSAGES/django.po:
```diff
@@ -259,79 +332,104 @@ msgstr "Etusivu"
-msgstr "Ei lupaa muokata."
+#: metarecord/views/classification.py:175 metarecord/views/function.py:251
+msgid "No permission for the state change."
+msgstr "Ei oikeutta tilan vaihtoon."
```
Nitpick-level 5/5, but I would use `lupaa` instead of `oikeutta` because that seems to be used in other texts when referring to permissions.

tuomas777 commented on metarecord/views/classification.py:
```diff
@@ -18,27 +21,59 @@ def include_related(request):
+        if user.has_perm(Classification.CAN_VIEW_MODIFIED_BY):
+            return obj.get_modified_by_display()
+
+        return None
```
Is this method actually needed, isn't `modified_by` deleted from the fields in `get_fields()` anyway when the user doesn't have the permission?

ehaivala replied:
Oh yeah. I added the deletion to `get_fields` afterwards and didn't come back to this. I'll fix this right up.

tuomas777 commented on metarecord/views/classification.py:
```diff
@@ -83,36 +118,166 @@ def _get_phases(self, obj):
+        if request.method == 'GET':
+            data = self._append_function_fields_to_repr(obj, data)
+
+            if request and not request.user.is_authenticated:
```
Checking for `request` to not be falsy should be probably on line 123 instead.

ehaivala commented:
Fixed the issues you commented about. Also I rebased this branch on top of latest `develop` and added some additional functionality that was requested when I demonstrated the changes to the users.

- Delete old unapproved classifications when newer approved version is saved.
- Add filtering by classification code an state to the admin list view
- Data migration to set all pre-existing classifications to approved

tuomas777 reviewed (approved):
LGTM, very nice work! 👍 🚀

ehaivala commented:
It was agreed that this pull request should be left open until there's an UI developer available to implement necessary changes for the UI to work with these backend changes. This can be merged when UI developer is starting to work on the changes.

ehaivala commented:
Got the permission to merge this now.
